File: src/main/java/net/runelite/client/plugins/oneclickagility/OneClickAgilityPlugin.java
package net.runelite.client.plugins.oneclickagility;

import eventbus.events.*;
import meteor.Main;
import meteor.util.GameEventManager;
import net.runelite.api.Client;
import net.runelite.api.DecorativeObject;
import net.runelite.api.GameObject;
import net.runelite.api.GameState;
import net.runelite.api.MenuAction;
import net.runelite.api.MenuEntry;
import net.runelite.api.Skill;
import net.runelite.api.Tile;
import net.runelite.api.TileItem;
import net.runelite.api.TileObject;
import net.runelite.api.coords.WorldPoint;
import net.runelite.api.widgets.Widget;
import net.runelite.api.widgets.WidgetID;
import net.runelite.api.widgets.WidgetInfo;
import net.runelite.api.widgets.WidgetItem;
import meteor.game.ItemManager;
import meteor.plugins.Plugin;
import meteor.plugins.PluginDescriptor;
import meteor.config.ConfigManager;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Set;

@PluginDescriptor(
        name = "One Click Agility",
        description = "Reclined gaming",
        enabledByDefault = false
)
public class OneClickAgilityPlugin extends Plugin
{

    private Client client  = Main.INSTANCE.getClient();

    GameEventManager gameEventManager = Main.INSTANCE.getGameEventManager();

    ItemManager itemManager = ItemManager.INSTANCE;

    private OneClickAgilityConfig config = (OneClickAgilityConfig) javaConfiguration(OneClickAgilityConfig.class);

    private ConfigManager configManager = ConfigManager.INSTANCE;

    private static final int MARK_ID = 11849;
    private static final int COIN_ID = 995;
    private static final Set<Integer> PORTAL_IDS = Set.of(36241,36242,36243,36244,36245,36246);
    private static final Set<Integer> SUMMER_PIE_ID = Set.of(7220,7218);
    private static final WorldPoint SEERS_END = new WorldPoint(2704,3464,0);
    private static final WorldPoint PYRAMID_TOP_RIGHT = new WorldPoint(3043,4697,3);
    private static final WorldPoint PYRAMID_TOP_LEFT = new WorldPoint(3042,4697,3);

    ArrayList<Tile> marks = new ArrayList<>();
    ArrayList<GameObject> portals = new ArrayList<>();
    DecorativeObject pyramidTopObstacle;
    GameObject pyramidTop;
    Course course;

    @Override
    public void onStart()
    {
        course = CourseFactory.build(config.courseSelection());
    }

    @Override
    public void onConfigChanged(ConfigChanged event)
    {
        if(event.getGroup().equals("oneclickagility"))
        {
            course = CourseFactory.build(config.courseSelection());
            gameEventManager.simulateGameEvents(this);
        }
    }


    @Override
    public void onMenuOptionClicked(MenuOptionClicked event)
    {
        if(event.getMenuOption().equals("<col=00ff00>One Click Agility"))
        {
            handleClick(event);
        }
        else if(event.getMenuOption().equals("One Click Agility"))
        {
            event.consume();
        }
    }

    @Override
    public void onClientTick(ClientTick event)
    {
        if(client.getLocalPlayer() == null || client.getGameState() != GameState.LOGGED_IN) return;
        String text;

        if(course.getCurrentObstacleArea(client.getLocalPlayer()) == null)
        {
            if (config.consumeMisclicks())
            {
                text = "One Click Agility";
            }
            else
            {
                return;
            }
        }
        else
        {
            text =  "<col=00ff00>One Click Agility";
        }

        client.insertMenuItem(
                text,
                "",
                MenuAction.UNKNOWN.getId(),
                0,
                0,
                0,
                true);
    }

    @Override
    public void onGameObjectSpawned(GameObjectSpawned event)
    {
        if(event.getGameObject() == null)
        {
            return;
        }

        if(event.getGameObject().getId() == 10869)
        {
            pyramidTop = event.getGameObject();
        }
        if (PORTAL_IDS.contains(event.getGameObject().getId()))
        {
            portals.add(event.getGameObject());
            return;
        }

        addToCourse(event.getGameObject());
    }

    @Override
    public void onGameObjectDespawned(GameObjectDespawned event)
    {
        if(event.getGameObject() == null)
        {
            return;
        }
        if (PORTAL_IDS.contains(event.getGameObject().getId()))
        {
            portals.remove(event.getGameObject());
            return;
        }
        if(event.getGameObject().getId() == 10869)
        {
            pyramidTop = null;
        }
        removeFromCourse(event.getGameObject());
    }

    @Override
    public void onWallObjectSpawned(WallObjectSpawned event)
    {
        addToCourse(event.getWallObject());
    }


    @Override
    public void onWallObjectDespawned(WallObjectDespawned event)
    {
        removeFromCourse(event.getWallObject());
    }


    @Override
    public void onDecorativeObjectSpawned(DecorativeObjectSpawned event)
    {
        if(event.getDecorativeObject().getId() == 10851)
        {
            if(pyramidTopObstacle == null || pyramidTopObstacle.getY() > event.getDecorativeObject().getY())
            {
                pyramidTopObstacle = event.getDecorativeObject();
                return;
            }
        }

        addToCourse(event.getDecorativeObject());
    }

    @Override
    public void onDecorativeObjectDespawned(DecorativeObjectDespawned event)
    {
        if(event.getDecorativeObject().getId() == 10851 && event.getDecorativeObject() == pyramidTopObstacle)
        {
            pyramidTopObstacle = null;
            return;
        }

        removeFromCourse(event.getDecorativeObject());
    }

    @Override
    public void onGroundObjectSpawned(GroundObjectSpawned event)
    {
        addToCourse(event.getGroundObject());
    }

    @Override
    public void onGroundObjectDespawned(GroundObjectDespawned event)
    {
        removeFromCourse(event.getGroundObject());
    }

    @Override
    public void onItemSpawned(ItemSpawned event)
    {
        if (event.getItem().getId() == MARK_ID)
        {
            marks.add(event.getTile());
        }
    }

    @Override
    public void onItemDespawned(ItemDespawned event)
    {
        if (event.getItem().getId() == MARK_ID)
        {
            marks.remove(event.getTile());
        }
    }

    private void addToCourse(TileObject tileObject)
    {
        if (course.obstacleIDs.contains(tileObject.getId()))
        {
            course.addObstacle(tileObject);
        }
    }

    private void removeFromCourse(TileObject tileObject)
    {
        if (course.obstacleIDs.contains(tileObject.getId()))
        {
            course.removeObstacle(tileObject);
        }
    }

    private void handleClick(MenuOptionClicked event)
    {
        if(config.skillBoost())
        {
            int boost = client.getBoostedSkillLevel(Skill.AGILITY)-client.getRealSkillLevel(Skill.AGILITY);
            if(config.boostAmount()>boost)
            {
                WidgetItem food = getWidgetItem(SUMMER_PIE_ID);
                if (food != null)
                {
                    event.setMenuEntry(createSummerPieMenuEntry(food));
                    return;
                }
            }
        }

        if(config.seersTele() && config.courseSelection() == AgilityCourse.SEERS_VILLAGE)
        {   //spellbook varbit, worldpoint of dropdown tile, teleportation animation ID
            if(client.getVarbitValue(4070) == 0 && client.getLocalPlayer().getWorldLocation().equals(SEERS_END) && client.getLocalPlayer().getAnimation() != 714)
            {
                event.setMenuEntry(createSeersTeleportMenuEntry());
                return;
            }
        }

        if(config.courseSelection() == AgilityCourse.AGILITY_PYRAMID)
        {
            if((client.getLocalPlayer().getWorldLocation().equals(PYRAMID_TOP_RIGHT) || client.getLocalPlayer().getWorldLocation().equals(PYRAMID_TOP_LEFT))
                    && pyramidTop.getRenderable().getModelHeight() == 309)
            {
                event.setMenuEntry(createPyramidTopMenuEntry());
                return;
            }
        }

        ObstacleArea obstacleArea = course.getCurrentObstacleArea(client.getLocalPlayer());
        if (obstacleArea == null)
        {
            return;
        }

        if (config.pickUpMarks() && !marks.isEmpty())
        {
            Tile wrongMarkTile = null;
            for (Tile mark : marks)
            {
                if (obstacleArea.containsObject(mark))
                {
                    Tile markTile = client.getScene().getTiles()[mark.getPlane()][mark.getSceneLocation().getX()][mark.getSceneLocation().getY()];

                    if (markTile != null && checkTileForMark(markTile))
                    {
                        if (config.pickUpCoins() && checkTileForCoins(markTile)) {
                            event.setMenuEntry(createCoinsMenuEntry(mark));
                        } else {
                            event.setMenuEntry(createMarkMenuEntry(mark));
                        }
                        return;
                    }
                    else
                    {
                        wrongMarkTile = mark;
                    }
                }
            }

            if(wrongMarkTile != null)
            {
                marks.remove(wrongMarkTile);
            }
        }

        if (!portals.isEmpty())
        {
            for(GameObject portal:portals)
            {
                if (obstacleArea.containsObject(portal) && portal.getClickbox() != null)
                {
                    event.setMenuEntry(createPortalMenuEntry(portal));
                    return;
                }
            }
        }
        if(config.consumeMisclicks() &&
                (client.getLocalPlayer().isMoving()
                        || client.getLocalPlayer().getPoseAnimation() != client.getLocalPlayer().getIdlePoseAnimation()))
        {
            event.consume();
            return;
        }

        event.setMenuEntry(obstacleArea.createMenuEntry());
    }

    private boolean checkTileForMark(Tile tile)
    {
        List<TileItem> items = tile.getGroundItems();
        if (items == null)
        {
            return false;
        }

        for (TileItem item:items)
        {
            if (item == null)
                continue;

            if(item.getId() == MARK_ID)
                return true;
        }
        return false;
    }

    private boolean checkTileForCoins(Tile tile)
    {
        List<TileItem> items = tile.getGroundItems();
        if (items == null)
        {
            return false;
        }

        for (TileItem item:items)
        {
            if (item == null)
                continue;

            if(item.getId() == COIN_ID)
                return true;
        }
        return false;
    }

    public WidgetItem getWidgetItem(Collection<Integer> ids) {
        Widget inventoryWidget = client.getWidget(WidgetInfo.INVENTORY);
        if (inventoryWidget != null) {
            Collection<WidgetItem> items = inventoryWidget.getWidgetItems();
            for (WidgetItem item : items) {
                if (ids.contains(item.getId())) {
                    return item;
                }
            }
        }
        return null;
    }

    private MenuEntry createSeersTeleportMenuEntry()
    {
        return client.createMenuEntry(
                "Seers'",
                "Camelot Teleport",
                2,
                MenuAction.CC_OP.getId(),
                -1,
                14286879,
                true);
    }

    private MenuEntry createSummerPieMenuEntry(WidgetItem food)
    {
        String[] foodMenuOptions = itemManager.getItemComposition(food.getId()).getInventoryActions();
        return client.createMenuEntry(
                foodMenuOptions[0],
                foodMenuOptions[0],
                food.getId(),
                MenuAction.ITEM_FIRST_OPTION.getId(),
                food.getIndex(),
                WidgetInfo.INVENTORY.getPackedId(),
                true);
    }

    private MenuEntry createMarkMenuEntry(Tile tile)
    {
        return client.createMenuEntry("Take",
                "Mark of Grace",
                MARK_ID,MenuAction.GROUND_ITEM_THIRD_OPTION.getId(),
                tile.getSceneLocation().getX(),
                tile.getSceneLocation().getY(),
                true);
    }

    private MenuEntry createCoinsMenuEntry(Tile tile)
    {
        return client.createMenuEntry("Take",
                "Coins",
                COIN_ID,MenuAction.GROUND_ITEM_THIRD_OPTION.getId(),
                tile.getSceneLocation().getX(),
                tile.getSceneLocation().getY(),
                true);
    }

    private MenuEntry createPortalMenuEntry(GameObject portal)
    {
        return client.createMenuEntry(
                "Travel",
                "Portal",
                portal.getId(),
                MenuAction.GAME_OBJECT_FIRST_OPTION.getId(),
                portal.getLocalLocation().getSceneX(),
                portal.getLocalLocation().getSceneY(),
                true
        );
    }

    private MenuEntry createPyramidTopMenuEntry()
    {
        return client.createMenuEntry(
                "Climb",
                "Climbing rocks",
                pyramidTopObstacle.getId(),
                MenuAction.GAME_OBJECT_FIRST_OPTION.getId(),
                pyramidTopObstacle.getLocalLocation().getSceneX(),
                pyramidTopObstacle.getLocalLocation().getSceneY(),
                true
        );
    }

}

File: client/src/main/kotlin/meteor/plugins/oneclickagility/OneClickAgilityPlugin.kt
package meteor.plugins.oneclickagility

import eventbus.events.*
import meteor.game.ItemManager
import meteor.plugins.Plugin
import meteor.plugins.PluginDescriptor
import meteor.util.GameEventManager
import net.runelite.api.*
import net.runelite.api.coords.WorldPoint
import net.runelite.api.widgets.WidgetInfo
import net.runelite.api.widgets.WidgetItem
import java.util.Set

@PluginDescriptor(name = "One Click Agility", description = "Reclined gaming", enabledByDefault = false)
class OneClickAgilityPlugin : Plugin() {

    var gameEventManager = GameEventManager
    var itemManager = ItemManager
    private val config = configuration<OneClickAgilityConfig>()

    var marks = ArrayList<Tile>()
    var portals = ArrayList<GameObject>()
    var pyramidTopObstacle: DecorativeObject? = null
    var pyramidTop: GameObject? = null
    var course: Course? = null
    override fun onStart() {
        resetCourse()
    }

    override fun onConfigChanged(it: ConfigChanged) {
        if (it.group == "oneclickagility") {
            resetCourse()
        }
    }

    fun resetCourse() {
        course = CourseFactory.build(config.courseSelection())
        gameEventManager.simulateGameEvents(this)
    }

    override fun onMenuOptionClicked(it: MenuOptionClicked) {
        if (it.getMenuOption() == "<col=00ff00>One Click Agility") {
            handleClick(it)
        } else if (it.getMenuOption() == "One Click Agility") {
            it.consume()
        }
    }

    override fun onClientTick(it: ClientTick) {
        if (client.localPlayer == null || client.gameState != GameState.LOGGED_IN) return
        val text: String = if (course!!.getCurrentObstacleArea(client.localPlayer) == null) {
            if (config.consumeMisclicks()) {
                "One Click Agility"
            } else {
                return
            }
        } else {
            "<col=00ff00>One Click Agility"
        }
        client.insertMenuItem(
            text,
            "",
            MenuAction.UNKNOWN.id,
            0,
            0,
            0,
            true
        )
    }

    override fun onGameObjectSpawned(it: GameObjectSpawned) {
        if (it.gameObject == null) {
            return
        }
        if (it.gameObject.id == 10869) {
            pyramidTop = it.gameObject
        }
        if (PORTAL_IDS.contains(it.gameObject.id)) {
            portals.add(it.gameObject)
            return
        }
        addToCourse(it.gameObject)
    }

    override fun onGameObjectDespawned(it: GameObjectDespawned) {
        if (it?.gameObject == null) {
            return
        }
        if (PORTAL_IDS.contains(it.gameObject.id)) {
            portals.remove(it.gameObject)
            return
        }
        if (it.gameObject.id == 10869) {
            pyramidTop = null
        }
        removeFromCourse(it.gameObject)
    }

    override fun onWallObjectSpawned(it: WallObjectSpawned) {
        addToCourse(it.wallObject)
    }

    override fun onWallObjectDespawned(it: WallObjectDespawned) {
        removeFromCourse(it.wallObject)
    }

    override fun onDecorativeObjectSpawned(it: DecorativeObjectSpawned) {
        if (it.decorativeObject.id == 10851) {
            if (pyramidTopObstacle == null || pyramidTopObstacle!!.y > it.decorativeObject.y) {
                pyramidTopObstacle = it.decorativeObject
                return
            }
        }
        addToCourse(it.decorativeObject)
    }

    override fun onDecorativeObjectDespawned(it: DecorativeObjectDespawned) {
        if (it.decorativeObject.id == 10851 && it.decorativeObject === pyramidTopObstacle) {
            pyramidTopObstacle = null
            return
        }
        removeFromCourse(it.decorativeObject)
    }

    override fun onGroundObjectSpawned(it: GroundObjectSpawned) {
        addToCourse(it.groundObject)
    }

    override fun onGroundObjectDespawned(it: GroundObjectDespawned) {
        removeFromCourse(it.groundObject)
    }

    override fun onItemSpawned(it: ItemSpawned) {
        if (it.item.id == MARK_ID) {
            marks.add(it.tile)
        }
    }

    override fun onItemDespawned(it: ItemDespawned) {
        if (it.item.id == MARK_ID) {
            marks.remove(it.tile)
        }
    }

    private fun addToCourse(tileObject: TileObject) {
        if (course!!.obstacleIDs.contains(tileObject.id)) {
            course!!.addObstacle(tileObject)
        }
    }

    private fun removeFromCourse(tileObject: TileObject) {
        if (course!!.obstacleIDs.contains(tileObject.id)) {
            course!!.removeObstacle(tileObject)
        }
    }

    private fun handleClick(it: MenuOptionClicked) {
        if (config.skillBoost()) {
            val boost = client.getBoostedSkillLevel(Skill.AGILITY) - client.getRealSkillLevel(Skill.AGILITY)
            if (config.boostAmount() > boost) {
                val food = getWidgetItem(SUMMER_PIE_ID)
                if (food != null) {
                    it.menuEntry = createSummerPieMenuEntry(food)
                    return
                }
            }
        }
        if (config.seersTele() && config.courseSelection() == AgilityCourse.SEERS_VILLAGE) {   //spellbook varbit, worldpoint of dropdown tile, teleportation animation ID
            if (client.getVarbitValue(4070) == 0 && client.localPlayer!!
                    .worldLocation == SEERS_END && client.localPlayer!!
                    .animation != 714
            ) {
                it.menuEntry = createSeersTeleportMenuEntry()
                return
            }
        }
        if (config.courseSelection() == AgilityCourse.AGILITY_PYRAMID) {
            if ((client.localPlayer!!
                    .worldLocation == PYRAMID_TOP_RIGHT || client.localPlayer!!
                    .worldLocation == PYRAMID_TOP_LEFT)
                && pyramidTop!!.renderable.modelHeight == 309
            ) {
                it.menuEntry = createPyramidTopMenuEntry()
                return
            }
        }
        val obstacleArea = course!!.getCurrentObstacleArea(client.localPlayer) ?: return
        if (config.pickUpMarks() && !marks.isEmpty()) {
            var wrongMarkTile: Tile? = null
            for (mark in marks) {
                if (obstacleArea.containsObject(mark)) {
                    val markTile = client.scene.tiles[mark.plane][mark.sceneLocation.x][mark.sceneLocation.y]
                    if (markTile != null && checkTileForMark(markTile)) {
                        if (config.pickUpCoins() && checkTileForCoins(markTile)) {
                            it.menuEntry = createCoinsMenuEntry(mark)
                        } else {
                            it.menuEntry = createMarkMenuEntry(mark)
                        }
                        return
                    } else {
                        wrongMarkTile = mark
                    }
                }
            }
            if (wrongMarkTile != null) {
                marks.remove(wrongMarkTile)
            }
        }
        if (portals.isNotEmpty()) {
            for (portal in portals) {
                if (obstacleArea.containsObject(portal) && portal.clickbox != null) {
                    it.menuEntry = createPortalMenuEntry(portal)
                    return
                }
            }
        }
        if (config.consumeMisclicks() &&
            (client.localPlayer!!.isMoving
                    || client.localPlayer!!.poseAnimation != client.localPlayer!!.idlePoseAnimation)
        ) {
            it.consume()
            return
        }
        obstacleArea.createMenuEntry()?.let {thisMenu->
            it.menuEntry = thisMenu
        }
    }

    private fun checkTileForMark(tile: Tile): Boolean {
        val items = tile.groundItems ?: return false
        for (item in items) {
            if (item == null) continue
            if (item.id == MARK_ID) return true
        }
        return false
    }

    private fun checkTileForCoins(tile: Tile): Boolean {
        val items = tile.groundItems ?: return false
        for (item in items) {
            if (item == null) continue
            if (item.id == COIN_ID) return true
        }
        return false
    }

    fun getWidgetItem(ids: Collection<Int?>): WidgetItem? {
        val inventoryWidget = client.getWidget(WidgetInfo.INVENTORY)
        if (inventoryWidget != null) {
            val items = inventoryWidget.widgetItems
            for (item in items) {
                if (ids.contains(item.id)) {
                    return item
                }
            }
        }
        return null
    }

    private fun createSeersTeleportMenuEntry(): MenuEntry {
        return client.createMenuEntry(
            "Seers'",
            "Camelot Teleport",
            2,
            MenuAction.CC_OP.id,
            -1,
            14286879,
            true
        )
    }

    private fun createSummerPieMenuEntry(food: WidgetItem): MenuEntry {
        val foodMenuOptions = itemManager.getItemComposition(food.id)!!.inventoryActions
        return client.createMenuEntry(
            foodMenuOptions[0],
            foodMenuOptions[0],
            food.id,
            MenuAction.ITEM_FIRST_OPTION.id,
            food.index,
            WidgetInfo.INVENTORY.packedId,
            true
        )
    }

    private fun createMarkMenuEntry(tile: Tile): MenuEntry {
        return client.createMenuEntry(
            "Take",
            "Mark of Grace",
            MARK_ID, MenuAction.GROUND_ITEM_THIRD_OPTION.id,
            tile.sceneLocation.x,
            tile.sceneLocation.y,
            true
        )
    }

    private fun createCoinsMenuEntry(tile: Tile): MenuEntry {
        return client.createMenuEntry(
            "Take",
            "Coins",
            COIN_ID, MenuAction.GROUND_ITEM_THIRD_OPTION.id,
            tile.sceneLocation.x,
            tile.sceneLocation.y,
            true
        )
    }

    private fun createPortalMenuEntry(portal: GameObject): MenuEntry {
        return client.createMenuEntry(
            "Travel",
            "Portal",
            portal.id,
            MenuAction.GAME_OBJECT_FIRST_OPTION.id,
            portal.localLocation.sceneX,
            portal.localLocation.sceneY,
            true
        )
    }

    private fun createPyramidTopMenuEntry(): MenuEntry {
        return client.createMenuEntry(
            "Climb",
            "Climbing rocks",
            pyramidTopObstacle!!.id,
            MenuAction.GAME_OBJECT_FIRST_OPTION.id,
            pyramidTopObstacle!!.localLocation.sceneX,
            pyramidTopObstacle!!.localLocation.sceneY,
            true
        )
    }

    companion object {
        private const val MARK_ID = 11849
        private const val COIN_ID = 995
        private val PORTAL_IDS = Set.of(36241, 36242, 36243, 36244, 36245, 36246)
        private val SUMMER_PIE_ID = Set.of(7220, 7218)
        private val SEERS_END = WorldPoint(2704, 3464, 0)
        private val PYRAMID_TOP_RIGHT = WorldPoint(3043, 4697, 3)
        private val PYRAMID_TOP_LEFT = WorldPoint(3042, 4697, 3)
    }
}